[CBRD-25244] Bad size for some RVDK_NEWVOL and RVDK_FORMAT logs and also bad string ending in disk_vhdr_set_vol_remarks

H2SU opened the pull request:
http://jira.cubrid.org/browse/CBRD-25244

Purpose

This issue covers two miscalculations in code.

- First one is when calculating the RVDK_NEWVOL and RVDK_FORMAT log size using disk_vhdr_length_of_varfields. 

- Second one is is disk_vhdr_set_vol_remarks, when length of vol_remarks is greater than maxsize. 

- Third one is executing memmove without trimming when vhdr size exceeds PAGESIZE using 'addvoldb' and 'renamedb' utilities leads to invading other memory regions.

Implementation


- Added a function to calculate the disk_volume_header size.

- Corrected an issue where memmove was performed without trimming when vhdr was expanded by renamedb or addvoldb, causing it to affect other memory regions.

- When adding remarks, fixed an issue where comparison of remarks length to usable_size did not account for the null character.


Remarks

- Added variables for readability.

H2SU committed:
[CBRD-25244] Fixed the code that incorrectly calculated the size of the disk_volume_header and added a new function to calculate its size.

H2SU committed:
[CBRD-25244] Fixed an issue where the null character was incorrectly placed when trimming long remarks.

H2SU committed:
[CBRD-25244] Executing memmove without trimming when vhdr length exceeds PAGESIZE using 'addvoldb' and 'renamedb' utilities leads to invading other memory regions.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -350,7 +350,7 @@ static bool disk_Logging = false;
 STATIC_INLINE char *disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr) __attribute__ ((ALWAYS_INLINE));
 STATIC_INLINE char *disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr) __attribute__ ((ALWAYS_INLINE));
-STATIC_INLINE int disk_vhdr_length_of_varfields (const DISK_VOLUME_HEADER * vhdr) __attribute__ ((ALWAYS_INLINE));
+STATIC_INLINE int disk_get_vhdr_size (const DISK_VOLUME_HEADER * vhdr) __attribute__ ((ALWAYS_INLINE));
```
볼륨 헤더와 관련된 함수는 `disk_vhdr_` prefix를 사용합니다. 일관성 있게 `disk_vhdr_get_size()` 또는 `disk_vhdr_get_vol_header_size()` 등을 사용하는게 어떨까요?

H2SU replied:
disk_vhdr까지가 prefix인지 몰라 위와 같은 이름을 사용하였습니다.
disk_vhdr_get_vol_header_size() 로 변경하여 반영하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5312,15 +5312,19 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+
+  remarks_size = (int) strlen (disk_vhdr_get_vol_remarks (vhdr)) + 1;
+
+  return (offsetof (DISK_VOLUME_HEADER, var_fields) + (vhdr->offset_to_vol_remarks + remarks_size));;
```
다음 방식이 더 심플한것 같은데, 어떠신가요?

```
char * remarks = disk_vhdr_get_vol_remarks (vhdr);

return (remarks + strlen (remarks) + 1) - ((char *) vhdr);
```

H2SU replied:
더 심플한 것 같습니다!
해당 내용 반영하여 커밋하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -527,16 +527,16 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
-  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH)
-      || (DB_PAGESIZE < (PGLENGTH) (SSIZEOF (DISK_VOLUME_HEADER) + strlen (vol_fullname) + 1)))
+  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH) ||
+      ((size_t) DB_PAGESIZE < ((size_t) (SSIZEOF (DISK_VOLUME_HEADER)) + strlen (vol_fullname) + 1)))
```
정확하게는 SSIZEOF (DISK_VOLUME_HEADER) 아닌 offsetof (DISK_VOLUME_HEADER, var_fields) 이겠네요.

H2SU replied:
기존 계산식 보다 그게 더 정확한 값이 맞기 때문에 변경하여 커밋하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -527,16 +527,16 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
 
   /* make sure that this is a valid purpose */
-  if (vol_purpose != DB_PERMANENT_DATA_PURPOSE && vol_purpose != DB_TEMPORARY_DATA_PURPOSE)
+  if ((vol_purpose != DB_PERMANENT_DATA_PURPOSE) && (vol_purpose != DB_TEMPORARY_DATA_PURPOSE))
```
이슈와 무관한 수정은 최소화 해주세요. 코드 가독성은 다소 주관적인 영역이며, 불필요한 수정으로 인해 리뷰 품질이 떨어질 수 있습니다.

H2SU replied:
넵. 추후 작업 시 해당 내용 반영하여 작업하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
  */
 static int
-disk_vhdr_set_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *vol_fullname)
+disk_vhdr_set_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *new_vol_fullname)
```
1. 이 함수가 old와 new를 모두 인자로 받고 있지 않기 때문에 new_ prefix는 불필요해 보입니다.
2. 함수 구현(define)의 인자명 수정 시 선언(declaration)의 인자명 역시 같이 변경해 주는게 좋습니다.

H2SU replied:
기존 vhdr에 있는 vol_fullname과 구분짓기 위해 사용하긴 했는데 사용하지 않는 것이 괜찮을까요?

hornetmj replied:
네이밍에 대한 부분보다는 다음을 생각해 주세요.
1. 본 이슈와 무관한 수정 최소화
2. 기존 vhdr에 있는 vol_fullname에 `cur_(current_)` 또는 `prev_(previous_)` 등을 붙여 구분함으로써 함수 내부로 수정 범위 최소화

H2SU replied:
아하 그러면 차라리 함수 내에서 변수명을 cur_vol_fullname 또는 prev_vol_fullname 으로 설정하고 인자로 받는 파라미터의 이름은 vol_fullname으로 받는 기존방식을 유지하는 쪽으로 변경해보겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  length_to_move = (length_diff + (int) strlen (vhdr->var_fields + length_diff) + 1
-		    - vhdr->offset_to_next_vol_fullname);
+  /* Contains null characters ( 1 byte ) */
+  vol_fullname_size = (int) strlen (vhdr->var_fields + vhdr->offset_to_vol_fullname) + 1;
+  next_vol_fullname_size = (int) strlen (vhdr->var_fields + vhdr->offset_to_next_vol_fullname) + 1;
+  remarks_size = (int) strlen (vhdr->var_fields + vhdr->offset_to_vol_remarks) + 1;
```
직접 접근하기 보다는 disk_vhdr_get_vol_fullname() 등 존재하는 inline 함수 사용을 권장합니다.
```
vol_fullname_size = (int) strlen (disk_vhdr_get_vol_fullname (vhdr)) + 1;
next_vol_fullname_size = (int) strlen (disk_vhdr_get_next_vol_fullname (vhdr)) + 1;
remarks_size = (int) strlen (disk_vhdr_get_vol_remarks (vhdr)) + 1;
```

H2SU replied:
위와 같이 수정하겠습니다!

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
-		 - (vhdr->offset_to_next_vol_fullname - vhdr->offset_to_vol_fullname));
+  name_length_diff = (new_vol_fullname_size - vol_fullname_size);
+
+  length_to_move = (next_vol_fullname_size + remarks_size);
```
다음 if 구문 내에서만 사용되기 때문에 if 구문 내에서 계산하면 되겠습니다.

H2SU replied:
수정하겠습니다!

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
+	  is_remarks_trim = true;
+
+	  length_to_move -= ((disk_get_vhdr_size (vhdr) + name_length_diff) - DB_PAGESIZE);
+	  remarks_size -= name_length_diff;
```
trim 되는 길이 만큼만 줄어드는게 맞지 않나요?

추가적으로 bool 타입의 is_remarks_trim 변수를 사용하는 것보다는 다음과 같이 int 타입의 trim_size 변수를 사용하는게 어떨까요?

```
int trim_size = disk_get_vhdr_size (vhdr) + name_length_diff - DB_PAGESIZE;

if (trim_size > 0)
{
    length_to_move -= trim_size;
    remarks_size -= trim_size;
}

...
```

hornetmj replied:
추가적으로 이 함수 기준으로는 반드시 remarks만 trim 되어 진다고 볼 수 없습니다. 따라서, is_`remarks`_trim 보다는 일반적인 trim_size로 제안드렸습니다.

H2SU replied:
trim 되는 길이 만큼만 줄어드는게 맞습니다!

잘못된 코드 수정하여 반영하고 추가적으로 rim_size로 변경하는 것이 가독성 면에서도 더 괜찮아보여 수정 후 커밋하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
+      memmove (disk_vhdr_get_next_vol_fullname (vhdr) + name_length_diff,
+	       disk_vhdr_get_next_vol_fullname (vhdr), length_to_move);
+
+      vhdr->offset_to_next_vol_fullname += name_length_diff;
+      vhdr->offset_to_vol_remarks += name_length_diff;
```
name_length_diff 값이 더해졌을 때 vhdr->offset_to_next_vol_fullname 및 vhdr->offset_to_vol_remarks가 page 영역을 벗어나면 안됩니다. 이를 고려한 코드가 되어야 합니다.

H2SU replied:
해당 내용 반영하여 수정해보겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5161,81 +5167,121 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  length_diff = vhdr->offset_to_vol_remarks;
-
-  length_to_move = (int) strlen (vhdr->var_fields + length_diff) + 1;
+  remarks_size = (int) strlen (vhdr->var_fields + vhdr->offset_to_vol_remarks) + 1;
+  next_vol_fullname_size = (int) strlen (vhdr->var_fields + vhdr->offset_to_next_vol_fullname) + 1;
```
disk_vhdr_get_next_vol_fullname() 와 disk_vhdr_get_vol_remarks() 를 사용하거나 코드 영역에서 자주 path 길이 등을 구하고 있는데, 이를 위한 함수를 소개하고 이를 이용하는게 어떨까요?
- disk_vhdr_get_vol_fullname_length()
- disk_vhdr_get_next_vol_fullname_length()
- disk_vhdr_get_vol_remarks_length()

H2SU replied:
그게 좀 더 보기 좋을 것 같습니다.

length() 가 아닌 size()가 더 적합할 것 같아 size()로 함수 추가해보겠습니다.

H2SU committed:
[CBRD-25244] Fixed code based on feedback

H2SU committed:
[CBRD-25244] Renamed parameter and local variable name in disk_vhdr_set_vol_fullname function

H2SU reviewed (commented):
피드백 내용 반영하여 커밋했습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5179,27 +5187,33 @@ disk_vhdr_set_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *new_vol_fulln
+  if (DB_PAGESIZE < (offsetof (DISK_VOLUME_HEADER, var_fields) + vhdr->offset_to_vol_remarks + name_length_diff))
+    {
+      return ER_BO_FULL_DATABASE_NAME_IS_TOO_LONG;
+    }
```
발생할 가능성이 희박하기 때문에 name_length_diff 이전에 명확한 사유(커멘트)와 함께 다음과 같은 assert를 추가하는 것으로 대체하는건 어떨까요? 의미는 어느 정도 분석해서 아시겠지만 boot_mount() 및 disk_set_creation(), disk_set_link() 함수 등에서 참조되는 vol_fullname과, next_vol_fullname가 잘리면(trim) 안됩니다.
```
assert (offsetof (DISK_VOLUME_HEADER, var_fields) + new_vol_fullname_size + next_vol_fullname_size + 1 <= DB_PAGESIZE);
```
좀 더 정확한 의미 전달을 위해 예를들어, disk_vhdr_is_trim_volname() 등의 함수나 매크로를 소개해 보는 것도 좋을 것 같네요.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5358,20 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+  char *remarks = disk_vhdr_get_vol_remarks (vhdr);
 
-  remarks_size = (int) strlen (disk_vhdr_get_vol_remarks (vhdr)) + 1;
+  if (vhdr == NULL)
```
선행 라인에서 vhdr이 이미 사용되고 있습니다. (disk_vhdr_get_vol_remarks 함수 인자)

추가적으로 0을 리턴하는 경우가 발생하면 안될 것 같습니다. assert() 처리 해주시면 어떨까요?
```
assert (vhdr != NULL);
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -516,27 +520,31 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
 
   /* make sure that this is a valid purpose */
-  if (vol_purpose != DB_PERMANENT_DATA_PURPOSE && vol_purpose != DB_TEMPORARY_DATA_PURPOSE)
+  if ((vol_purpose != DB_PERMANENT_DATA_PURPOSE) && (vol_purpose != DB_TEMPORARY_DATA_PURPOSE))
```
```suggestion
  if (vol_purpose != DB_PERMANENT_DATA_PURPOSE && vol_purpose != DB_TEMPORARY_DATA_PURPOSE)
```

hornetmj replied:
원복 요청 - 불필요한 코드 수정은 코드 라인의 수정 이력을 바꾸기 때문에 최소화 해주시는게 좋습니다. (git blame 참고)

H2SU replied:
원복하여 반영하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -545,16 +553,16 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
   /* safe guard: permanent volumes with temporary data purpose must be maxed */
-  assert (vol_purpose != DB_TEMPORARY_DATA_PURPOSE || ext_info->voltype != DB_PERMANENT_VOLTYPE
-	  || ext_info->nsect_total == ext_info->nsect_max);
+  assert ((vol_purpose != DB_TEMPORARY_DATA_PURPOSE) || (ext_info->voltype != DB_PERMANENT_VOLTYPE)
+	  || (ext_info->nsect_total == ext_info->nsect_max));
```
```suggestion
  assert (vol_purpose != DB_TEMPORARY_DATA_PURPOSE || ext_info->voltype != DB_PERMANENT_VOLTYPE
	  || ext_info->nsect_total == ext_info->nsect_max);
```

H2SU replied:
원복하여 반영하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
  */
 static int
-disk_vhdr_set_next_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *next_vol_fullname)
+disk_vhdr_set_next_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *new_next_vol_fullname)
```
```suggestion
disk_vhdr_set_next_vol_fullname (DISK_VOLUME_HEADER * vhdr, const char *next_vol_fullname)
```
주석 및 관련 코드 같이 원복 검토

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
   if (vol_remarks != NULL)
     {
-      maxsize = (DB_PAGESIZE - offsetof (DISK_VOLUME_HEADER, var_fields) - vhdr->offset_to_vol_remarks);
+      usable_size = (DB_PAGESIZE - offsetof (DISK_VOLUME_HEADER, var_fields) - vhdr->offset_to_vol_remarks);
```
NIT: (중요하지는 않지만 더 나은 방법이 있다는 의견)

수식을 바로 쓰는 경우 추후 코드 가독성이 떨어질 수 있습니다. 위 수식은 DISK_VOLUME_HEADER의 구조를 알고 있어야 이해가 되는 코드입니다. disk_vhdr_get_free_space() 등의 함수 소개를 통해 좀 더 명확한 의미 전달을 하는게 어떨까요?

필요하다면 고정 길이 필드와 가변 길이 필드를 구분해서 함수를 소개하고, 상위에서 이를 호출하거나 이를 이용한 상위 함수를 소개하는 방법 역시도 코드를 깔끔하게 정리할 수 있는 방법 중 하나일 것 입니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+
 /*
- * disk_vhdr_length_of_varfields () - get length of volume header including variable fields
+ * disk_vhdr_get_vol_header_size () - get size of volume header including variable fields
```
```suggestion
 * disk_vhdr_get_vol_header_size () - get the size of volume header including variable fields
```

H2SU replied:
해당 부분 수정하여 반영하겠습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
 {
-  return (vhdr->offset_to_vol_remarks + (int) strlen (disk_vhdr_get_vol_remarks (vhdr)));
+  char *remarks = disk_vhdr_get_vol_remarks (vhdr);
+
```
```suggestion
  assert (vhdr != nullptr);
  
  const char *remarks = disk_vhdr_get_vol_remarks (vhdr);

```

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+ * disk_vhdr_get_vol_header_size () - get size of volume header including variable fields
  *
- * return    : total length
+ * return    : total size
```
length 에서 size로 변경한 이유가 있을까요?

H2SU replied:
strlen과 같은 length를 반환하는 함수는 null문자는 포함하지 않은 길이만 반환하게 되는데, 해당 함수는 null문자가 들어가있는 공간까지 포함하여 반환해주기 때문에 size라는 표현이 더 맞지 않을까 생각해서 변경하였습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+}
+
+/*
+ * disk_vhdr_get_vol_remarks_size () - get remarks_size from volume header
```
```suggestion
 * disk_vhdr_get_vol_remarks_size () - get the size of 'remarks' from volume header
```

H2SU replied:
반영하겠습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+/*
+ * disk_vhdr_get_vol_remarks_size () - get remarks_size from volume header
+ *
+ * return    : remarks size
```
```suggestion
 * return    : the size of remarks
```

H2SU replied:
반영하겠습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+ * return    : remarks size
+ * vhdr (in) : volume header
+ */
+STATIC_INLINE int
```
size_t가 아닌 int형을 반환해주는 이유가 있을까요?

H2SU replied:
그 부분에 대해서 따로 크게 생각하고 작성한 건 아니였습니다.
말씀해주셔서 한 번 찾아봤는데 size_t로 반환하는게 더 적합한 부분인 것 같아 수정해보겠습니다.

H2SU replied:
해당 함수를 호출하여 리턴 값을 저장하는 변수의 형이 int형이라 int형을 반환하는 것으로 결정하였습니다!

joohok replied:
리턴 값을 저장하는 변수의 데이터 타입은 변경하면 문제가 생길까요? 
size가 0보다 작아질 경우가 없을 것 같아서 말씀드립니다.

H2SU replied:
현재 이름보다 변경하려는 이름이 더 짧은 경우 diff값을 나타내는 name_length_diff값이 음수일 경우가 존재합니다.
때문에 name_length_diff는 unsigned int가 아닌 int로 선언해야하기 때문에
이를 계산할 때 사용하는 변수를 int형으로 선언하였습니다!
size_t 로 선언을 하고 계산할 때만 (int)로 형변환을 하는 방식이 더 좋은 방법인가요?

joohok replied:
특정 기능을 별도의 함수로 모듈화 했다면, 이 함수가 어디서든 호출 될 수 있다는 것을 가정해야합니다. 
특정 caller에서 사용되는 형태를 기준으로 return type을 정한다는 게 어색합니다. 
값을 반환할때마다 강제로 형변환 하는데, 그 과정에서 어떤 의도치않은 변환이 일어날지 알 수 없고요.

H2SU replied:
size_t로 반환값을 설정하고 함수를 호출하는 부분에서 int로 형변환 하는 코드로 변경하였습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5374,62 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
 
+
+/*
+ * disk_vhdr_get_vol_fullname_size () - get vol_fullname_size from volume header
```
function description을 작성할 것이라면, 이 함수가 무엇을 하는지 직관적으로 보이면 좋을 것 같습니다. 
volume header에는 vol_fullname_size 라는 변수는 존재하지 않습니다. 
```suggestion
 * disk_vhdr_get_vol_fullname_size () - get the size of the volume's fullname using the vhdr(in).
```

H2SU replied:
해당 사항 반영하여 수정하도록 하겠습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
- *   next_vol_fullname(in):
+ *   return: 
+ *   vhdr(in): disk volume header
+ *   new_next_vol_fullname(in): new next volume full name
```
```suggestion
 *   new_next_vol_fullname(in): full name of the next volume to be set
```

H2SU replied:
해당 사항 반영하여 수정하도록 하겠습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  length_to_move = (length_diff + (int) strlen (vhdr->var_fields + length_diff) + 1
-		    - vhdr->offset_to_next_vol_fullname);
+  /* Contains null characters ( 1 byte ) */
+  old_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
+  next_vol_fullname_size = disk_vhdr_get_next_vol_fullname_size (vhdr);
+  remarks_size = disk_vhdr_get_vol_remarks_size (vhdr);
+  new_vol_fullname_size = (int) strlen (vol_fullname) + 1;
```
```suggestion
  const int old_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
  const int next_vol_fullname_size = disk_vhdr_get_next_vol_fullname_size (vhdr);
  const int remarks_size = disk_vhdr_get_vol_remarks_size (vhdr);
  const int vol_fullname_size = (int) strlen (vol_fullname) + 1;
```

H2SU replied:
const는 생각도 못 했던 부분인데 상수값이라는 점을 알리는 면이나 추후 발생할 수 있는 에러 방지면에서 좋은 방법인 것 같아 수정해보겠습니다.

joohok replied:
희수님의 스타일에 따라 변경하시면 될 것 같습니다

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  int length_to_move;
+  int name_length_diff = 0;
+  int length_to_move = 0;
+  int old_vol_fullname_size = 0;
+  int next_vol_fullname_size = 0;
+  int remarks_size = 0;
+  int new_vol_fullname_size = 0;
```
```suggestion
```

H2SU replied:
어떤 부분이 바뀐 건지 알 수 있을까요..?

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
   /* Difference in length between new name and old name */
-  length_diff = (((int) strlen (vol_fullname) + 1)
-		 - (vhdr->offset_to_next_vol_fullname - vhdr->offset_to_vol_fullname));
+  name_length_diff = (new_vol_fullname_size - old_vol_fullname_size);
```
```suggestion
  name_length_diff = (vol_fullname_size - old_vol_fullname_size);
```

H2SU replied:
한 눈에 보기 쉽게 new , old를 사용하긴 했는데 vol_fullname_size 라는 명칭이 더 적합할까요..?

joohok replied:
new_vol_fullname_size 를 구하기 위해 사용되는 변수가 vol_fullname 이어서 제안드렸습니다. 

개인적으로는 new 와 old 둘다 명시할 필요가 없을거라 생각합니다만, 사소한 것이기 때문에 희수님 판단에 따라 결정하시면 될 것 같네요.

H2SU replied:
new old prefix를 사용하는게 좀 더 보기 편한 것 같아 사용하는 것으로 결정해보겠습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
 
-  if (length_diff != 0)
+  /* When the new_vol_fullname is too long, causing other var_fields to overflow beyond the page. */
+  if (DB_PAGESIZE < (offsetof (DISK_VOLUME_HEADER, var_fields) + vhdr->offset_to_vol_remarks + name_length_diff))
```
Nit.
추후 코드 유지보수를 위해 이 조건이 무엇을 확인하기 위한 것인지 의미를 가지는 단위(Inline 함수 또는 매크로)로 대체하면 좋을 것 같습니다.

H2SU replied:
한 번 어떻게 대체하면 좋을지 고민해보겠습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
+  if (name_length_diff != 0)
     {
+      /* length to move */
+      remarks_size = disk_vhdr_get_vol_remarks_size (vhdr);
```
```suggestion
      const int remarks_size = disk_vhdr_get_vol_remarks_size (vhdr);
```

H2SU replied:
remarks size는 상수가 아니기때문에 int로 하는 것이 맞을 것 같습니다.

joohok replied:
수정 예정인가요?

H2SU replied:
수정하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  int next_vol_fullname_size;
+  int next_vol_fullname_size = 0;
+  int new_next_vol_fullname_size = 0;
+  int remarks_size = 0;
```
```suggestion
```

H2SU replied:
어떤 의미인지 이해를 못 했습니다ㅠ

joohok replied:
파일 내의 리뷰들을 다 봐주시길 바랍니다. 
해당 변수를 아래 5253L 으로 옮기면, 이 scope에서는 사용되지 않는 변수입니다.

H2SU replied:
수정하였습니다.

H2SU committed:
[CBRD-25244] Modify parameter and local variable names in disk_vhdr_set_next_vol_fullname function.

joohok reviewed (commented):
아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. 
disk_set_creation ()
disk_set_link ()

H2SU committed:
[CBRD-25244] Revert unnecessary changes and fix function description comments.

H2SU committed:
[CBRD-25244] Change some variables storing size to 'const'.

H2SU committed:
[CBRD-25244] Change the return type of the function to obtain size to size_t

H2SU committed:
[CBRD-25244] Defined the maximum length of the variable field to prevent trimming of remarks.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,12 @@ static DKNSECTS disk_Temp_max_sects = -2;
+#define DISK_VOLUME_MAX_PATH_LENGTH DB_MAX_PATH_LENGTH
+#define DISK_VOLUME_REMARKS_MAX_LENGTH (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DISK_VOLUME_MAX_PATH_LENGTH + DISK_VOLUME_MAX_PATH_LENGTH))
+
+
+
```
```suggestion
```

joohok replied:
불필요한 blank는 제거해도 좋을 것 같습니다.

H2SU replied:
넵 수정하겠습니다.

joohok replied:
반영되지 않았습니다.

H2SU replied:
수정했었는데 이후 코드를 건들면서 생긴 것 같습니다. 다시 수정하여 제거했습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5177,106 @@ disk_type_to_string (DB_VOLTYPE voltype)
 
-  length_to_move = (int) strlen (vhdr->var_fields + length_diff) + 1;
+  const int old_next_vol_fullname_size = disk_vhdr_get_next_vol_fullname_size (vhdr);
 
   /* Difference in length between new name and old name */
-  length_diff = (next_vol_fullname_size - (vhdr->offset_to_vol_remarks - vhdr->offset_to_next_vol_fullname));
+  const int name_length_diff = (new_next_vol_fullname_size - old_next_vol_fullname_size);
```
제 생각엔 이 부분은 C언어 스타일이 아닌 것 같네요. 변수 또는 상수들은 Scope 내에서 가장 먼저 정의되어야할 것 같습니다. 
정의하는 시점에 미리 알 수 있고, 정해진 값이라면 const를 사용하는게 좋아보입니다.

H2SU replied:
넵 수정하겠습니다.

H2SU replied:
수정하여 커밋하였습니다.

joohok commented:
> 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()

답변 부탁드립니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5177,106 @@ disk_type_to_string (DB_VOLTYPE voltype)
 
-  length_to_move = (length_diff + (int) strlen (vhdr->var_fields + length_diff) + 1
-		    - vhdr->offset_to_next_vol_fullname);
+  assert (new_vol_fullname_size <= DISK_VOLUME_MAX_PATH_LENGTH);
```
DISK_VOLUME_MAX_PATH_LENGTH 보다 큰 경우에는 어떻게 처리될까요?

H2SU replied:
MAX_LENGTH를 초과하면 assert를 발생시키려고 위와 같은 코드를 작성하였는데, 에러코드를 넣어 에러를 출력하는 방향으로 가는게 좋을까요?

joohok replied:
new_vol_fullname_size 가 별다른 처리 없이도 반드시 DISK_VOLUME_MAX_PATH_LENGTH 값보다 작은가요? 
assert는 예외처리 함수가 아닙니다. 반드시 지켜져야하지만, 모든 상황을 체크할 수 없기 때문에 넣는 디버깅용 체크 함수입니다.

joohok replied:
Caller에서 예외처리가 되기 때문에 assertion으로 충분하다는 답변 받았습니다.

H2SU replied:
disk_vhdr_set_vol_fullname 함수와 disk_vhdr_set_next_vol_fullname 함수 호출 시
해당 함수 외부에서 volume path가 너무 길면 예외처리하여 에러가 발생하도록 설정되어있어 위 두 함수 내부로는 절대 MAX_PATH_LENGTH보다 큰 값이 들어올 수 없기 때문에 assert로 처리해도 될 것 같습니다!

joohok reviewed (commented):
Assertion 보다는 예외처리가 필요할 것 같습니다.

H2SU commented:
> > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> 
> 답변 부탁드립니다.

예외처리는 함수 내부에서 진행하기 때문에
따로 수정은 필요 없을 것 같습니다.

H2SU committed:
[CBRD-25244] Changed constant to a variable.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5325,59 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+}
+
+/*
+ * disk_vhdr_get_next_vol_fullname_size () - get next_vol_fullname_size from volume header
```
의미를 명확하게 해주세요. 동일한 리뷰가 반복되고 있습니다.

joohok replied:
모든 리뷰에 대응해주세요. 매 리뷰 라운드마다 응답이 없는 리뷰가 존재합니다.

H2SU replied:
어떤 부분의 의미를 명확하게 해야하는건지 이해하지 못 하였습니다.
함수에 대한 설명이 명확하지 않은 것인가요..?

H2SU replied:
수정하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5325,59 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+{
+  assert (vhdr != NULL);
+
+  char *remarks = disk_vhdr_get_vol_remarks (vhdr);
```
```suggestion
  const char *remarks = disk_vhdr_get_vol_remarks (vhdr);
```

H2SU replied:
수정하였습니다.

joohok reviewed (commented):
DISK_VOLUME_MAX_PATH_LENGTH 보다 큰 경우에 대한 처리가 없습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5177,109 @@ disk_type_to_string (DB_VOLTYPE voltype)
     {
-      maxsize = (DB_PAGESIZE - offsetof (DISK_VOLUME_HEADER, var_fields) - vhdr->offset_to_vol_remarks);
+      const int remarks_size = (int) (strlen (vol_remarks) + 1);
+      assert (remarks_size <= DISK_VOLUME_REMARKS_MAX_LENGTH);
```
remark 사이즈에 대한 예외도 caller에서 체크될까요?

H2SU replied:
체크되지 않고 있어 disk_vhdr_set_vol_remarks 함수 내에서 예외처리 해야할 것 같습니다.

H2SU committed:
[CBRD-25244] change variable to const constant

H2SU committed:
[CBRD-25244] Add ER_DISK_REMARKS_LENGTH_TOO_LONG error and handle error when REMARKS length exceeds MAX_LENGTH.

joohok commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 Remarks length (%1$d) is too long; it must not exceed (%2$d) bytes.
```
```suggestion
1361 Remarks length (%1$d) is too long; it must not exceed %2$d bytes.
```

H2SU replied:
변경하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,10 @@ static DKNSECTS disk_Temp_max_sects = -2;
 #define DISK_STAB_UNIT_SIZE_OF sizeof (DISK_STAB_UNIT)
 
+/* The maximum size of the variable field in the disk volume header. */
+#define DISK_VOLUME_MAX_PATH_LENGTH DB_MAX_PATH_LENGTH
```
기존의 DB_MAX_PATH_LENGTH를 사용하지 않고, 별도로 정의한 이유가 있을까요?
또한 size와 length가 혼용되고 있어서 이해하기 힘듭니다. 
CUBRID 코드들을 참고하여 기준을 잡고 용어를 사용했으면 좋겠습니다.

H2SU replied:
기존의 DB_MAX_PATH_LENGTH를 사용할까 고민을 하긴했는데, DISK prefix를 사용하는게 일관성이 있어보여 위와같이 정의하였습니다.
DB_MAX_PATH_LENGTH를 사용하는 것이 더 나을까요?
length는 해당 문자열의 길이라는 뜻으로 사용하려했고, size는 해당 문자열이 그래서 메모리의 얼마만큼을 사용하는가?의 뜻으로 사용하려고 둘을 나눠사용하였습니다.
CUBRID코드에서도 뭔가 일관성있게 사용되는 것 같아 보이지 않아 어느 코드를 메인으로 참고하여 따라가야할지 모르겠습니다...ㅠ

H2SU replied:
기존 DB_MAX_PATH_LENGTH를 사용하는 것으로 수정하였습니다.

YeunjunLee commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +524,34 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
-  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH)
-      || (DB_PAGESIZE < (PGLENGTH) (SSIZEOF (DISK_VOLUME_HEADER) + strlen (vol_fullname) + 1)))
+  if ((vol_fullname_size > DB_MAX_PATH_LENGTH) ||
+      ((offsetof (DISK_VOLUME_HEADER, var_fields) + vol_fullname_size) > DB_PAGESIZE))
```
`vol_fullname_size`를 `DB_MAX_PATH_LENGTH`로 제한하면, 조건문의 두 번째 조건 이제 확인하지 않아도 될 것 같습니다.

H2SU replied:
vol_fullname_size가 DB_MAX_PATH_LENGTH를 넘어가진 않지만 vhdr와 크기를 더했을 때 PAGESIZE를 초과하는 케이스가 존재합니다.

예 ) vhdr(124) + vol_fullname_size(4095) > DB_PAGESIZE(4K~16K)

위 조건문은 추후 DB_PAGESIZE에 따라 조건문을 다르게 수정할 계획에 있습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,10 @@ static DKNSECTS disk_Temp_max_sects = -2;
 
+/* The maximum size of the variable field in the disk volume header. */
+#define DISK_VOLUME_MAX_PATH_LENGTH DB_MAX_PATH_LENGTH
+#define DISK_VOLUME_REMARKS_MAX_LENGTH (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DISK_VOLUME_MAX_PATH_LENGTH + DISK_VOLUME_MAX_PATH_LENGTH))
```
DISK_VOLUME_REMARKS_MAX_LENGTH 값이 무엇인지, 이 값을 정의한 이유가 무엇인지 코드를 통해 알 수가 없습니다. 
따라서 이 값이 무엇인지, 왜 정의했는지 등을 comment로 남기면 좋을 것 같습니다. 

1. 수식 설명 필요 (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DISK_VOLUME_MAX_PATH_LENGTH + DISK_VOLUME_MAX_PATH_LENGTH))
2. 기존에는 없던 값인데, 새롭게 정의한 이유 설명

H2SU replied:
수정하였습니다.

joohok commented:
> > > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> > 
> > 
> > 답변 부탁드립니다.
> 
> 예외처리는 함수 내부에서 진행하기 때문에 따로 수정은 필요 없을 것 같습니다.

어떤 예외처리가 되어있는걸까요? 
위 함수들에서 volume name 길이를 다룹니다. 이 이슈에서 volume name에 대해 logging 하는 부분들도 처리하길래 질문드렸습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,108 +5173,157 @@ disk_type_to_string (DB_VOLTYPE voltype)
-		    - vhdr->offset_to_next_vol_fullname);
+  /* Contains null characters ( 1 byte ) */
+  old_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
+  next_vol_fullname_size = disk_vhdr_get_next_vol_fullname_size (vhdr);
```
vhdr가 NULL 일 경우에 대해 예외처리가 필요할 것 같습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5182,108 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  length_diff = vhdr->offset_to_vol_remarks;
+  /* Contains null characters ( 1 byte ) */
+  const int new_vol_fullname_size = (int) strlen (vol_fullname) + 1;
+  const int old_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
```
vhdr이 null인지 체크해야할 것 같네요. vhdr을 인자로 받는 모든 함수들에서 예외처리 또는 assert가 필요해보입니다.

H2SU replied:
disk_vhdr_set_vol_fullname, disk_vhdr_set_next_vol_fullname, disk_vhdr_set_vol_remarks 함수를 호출하는 곳에서 vhdr이 null로 들어오는 경우가 발생하지 않아서 해당 함수 내에서 assert로 혹시라도 다른 곳에서 해당 함수를 사용할 때 NULL을 넣는 경우에 오류를 내도록 처리만 하여 수정하였습니다.

H2SU committed:
[CBRD-25244] Add comments and assertions, and fix error message

H2SU commented:
> > > > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> > > 
> > > 
> > > 답변 부탁드립니다.
> > 
> > 
> > 예외처리는 함수 내부에서 진행하기 때문에 따로 수정은 필요 없을 것 같습니다.
> 
> 어떤 예외처리가 되어있는걸까요? 위 함수들에서 volume name 길이를 다룹니다. 이 이슈에서 volume name에 대해 logging 하는 부분들도 처리하길래 질문드렸습니다.

H2SU commented:
> > > > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> > > 
> > > 
> > > 답변 부탁드립니다.
> > 
> > 
> > 예외처리는 함수 내부에서 진행하기 때문에 따로 수정은 필요 없을 것 같습니다.
> 
> 어떤 예외처리가 되어있는걸까요? 위 함수들에서 volume name 길이를 다룹니다. 이 이슈에서 volume name에 대해 logging 하는 부분들도 처리하길래 질문드렸습니다.

disk_set_creation() 함수는 DB_MAX_PATH_LENGTH로 에러처리를 하고 있는데 disk_set_link() 함수는 해당 함수를 호출하는 함수에서 에러처리를 하고 있는 부분도 있고 아닌 부분도 존재하고 있습니다.

disk_set_creation() 함수 내부에서 DB_MAX_PATH_LENGTH로 에러처리를 해주는 것이 좋아보여 에러처리 코드 추가하려합니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,14 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ * To address this, we have removed the trimming functionality and defined a maximum length for remarks. 
+ * Now, remarks longer than this defined maximum length will be restricted from being entered. 
+ */
+#define DISK_VOLUME_REMARKS_MAX_LENGTH (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DB_MAX_PATH_LENGTH + DB_MAX_PATH_LENGTH))
```
```suggestion
#define DISK_VOLUME_HEADER_REMARKS_MAX_SIZE (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DB_MAX_PATH_LENGTH + DB_MAX_PATH_LENGTH))
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +528,34 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
+      er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_DISK_REMARKS_LENGTH_TOO_LONG, 2, NULL, remarks_size,
+	      DISK_VOLUME_REMARKS_MAX_LENGTH);
+      return ER_DISK_REMARKS_LENGTH_TOO_LONG;
+    }
```
```suggestion
    }

```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5186,114 @@ disk_type_to_string (DB_VOLTYPE voltype)
 
-  length_to_move = (length_diff + (int) strlen (vhdr->var_fields + length_diff) + 1
-		    - vhdr->offset_to_next_vol_fullname);
+  /* Contains null characters ( 1 byte ) */
+  const int new_vol_fullname_size = (int) strlen (vol_fullname) + 1;
+  const int old_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
```
```suggestion
  const int vol_fullname_size = (int) strlen (vol_fullname) + 1;
  const int cur_vol_fullname_size = disk_vhdr_get_vol_fullname_size (vhdr);
```

hornetmj replied:
추가적으로 상수화(const)가 굳이 필요하지 않다고 생각되고, 이를 위해 assert (vhdr != NULL) 체크가 변수 선언부 중간에 삽입되어 있는 것도 어색하게 느껴집니다. 다른 코드부에서도 지역 변수들에 사용된 const 키워드에 대해 제거 검토해 주세요.

H2SU replied:
const 사용을 언제 하는 것이 좋고 필요없는 순간은 언제인지 잘 모르겠습니다. 수정된 코드 재리뷰 부탁드립니다.

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5186,114 @@ disk_type_to_string (DB_VOLTYPE voltype)
 {
-  int maxsize;
   int ret = NO_ERROR;
+  const int remarks_size = (int) (strlen (vol_remarks) + 1);
```
vol_remarks가 NULL인 경우 strlen()는 처리하지 못합니다. 아래 5294 라인에서 vol_remarks에 대한 NULL 검사를 수행하고 있습니다. 따라서, 이를 고려해서 strlen(vol_remarks)를 수행할 위치를 조정해야 합니다.

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5339,59 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
  */
-STATIC_INLINE int
-disk_vhdr_length_of_varfields (const DISK_VOLUME_HEADER * vhdr)
+STATIC_INLINE size_t
```
이 함수 리턴값을 받는 변수들이 int 형식입니다. size_t 형식 보다는 int 형식으로 리턴하는게 어떨까요?

H2SU replied:
수정하였습니다.

hornetmj replied:
반영되어 있지 않습니다. :)

H2SU replied:
해당 내용을 수정하였으나, 주호님이 특정 기능을 별도의 함수로 모듈화 했다면, 이 함수가 어디서든 호출 될 수 있다는 것을 가정하여 호출하는 곳에서 형변환을 해서 사용하는 것이 맞는 것 같다하셔서 다시 size_t로 수정하였습니다.

hornetmj replied:
PR 교육 과정에서 설명드린 내용이기에 질문드립니다. 이슈 담당자는 납득되지 않는 방식/의견을 수용하면 안된다고 설명드렸습니다. 수렴하신 내용에 대해 충분히 납득하시고 동의하신건가요? 동의하셨다면, 어떤 근거/이점을 들어 최종 선택하신 방법이 맞다라고 이슈 담당자 관점에서 설명하실 수 있을까요? 추후 이슈 처리에서 중요한 부분이기에 질문드립니다.

H2SU replied:
주호님이 말씀해주신 부분을 듣고 납득이 되어 수정하였습니다.
현재는 함수를 호출하여 값을 저장하는 변수들이 전부 int형이지만, 
함수를 만들 땐 해당 함수가 추후 언제 어디서든 불릴 수 있다는 것을 전제하에 만들어야하기 때문에
최대한 함수의 목적에 맞게 작성하는 것이 맞는 것 같다는 생각이 들었습니다.
그 부분에 납득이 되어 수정하였습니다.

H2SU committed:
[CBRD-25244] Fix variable field length handling in disk volume header

- Renamed DISK_VOLUME_REMARKS_MAX_LENGTH to DISK_VOLUME_HEADER_REMARKS_MAX_SIZE for clarity.
- Changed the return types of several functions to  from  to ensure consistent signedness.
- Added assertions and casts to address warnings about signed/unsigned integer comparisons.
- Ensured that variable field sizes (volume fullname, next volume fullname, and remarks) adhere to the defined maximum lengths.
- Updated comments to reflect changes and clarify the structure of the DISK_VOLUME_HEADER.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +532,35 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
       return ER_BO_FULL_DATABASE_NAME_IS_TOO_LONG;
     }
 
+  if (remarks_size > (int) DISK_VOLUME_HEADER_REMARKS_MAX_SIZE)
```
매크로 형변환 관련
- 컴파일 워닝 제거가 목적이시라면 형변환 자체를 매크로 정의로 옮겨주시는게 가독성 관점에서 좋을 것 같습니다.
- 컴파일 워닝이 발생하지 않는다면 명시적 형변환을 수행할 필요는 없을 것 같습니다.

H2SU replied:
서로 다른 형끼리 비교를 하는게 추후 문제가 될만한 상황은 없을까요?
당장 컴파일 워닝이 발생하진 않습니다!

H2SU replied:
매크로 선언 위치에 형변환하는 것으로 수정하였습니다.

hornetmj commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 Remarks length (%1$d) is too long; it must not exceed %2$d bytes.
```
사용자는 createdb나 addvoldb 유틸리티에서 --comment 옵션을 사용합니다. Remarks 라는 용어로는 쉽게 원인 파악을 못할 수 있습니다. 'createdb 또는 addvoldb의 --comment 옵션으로 전달된 주석의 길이가 x 바이트를 초과했습니다.' 정도의 의미로 작성해 보시면 어떨까요? 참고로 '주석(comment)'이란 용어는 한글 매뉴얼에서 참고된 용어입니다.

H2SU replied:
수정하였습니다!

H2SU committed:
[CBRD-25244] fix the error message (Remarks->Comments)

H2SU committed:
[CBRD-25244] Delete type conversion

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5201,120 @@ disk_type_to_string (DB_VOLTYPE voltype)
+  int remarks_size = 0;
 
-  length_diff = vhdr->offset_to_vol_remarks;
+  assert (vhdr != NULL);
```
```suggestion
  assert (vhdr != NULL);
  assert (vol_fullname != NULL);
```
nit

H2SU replied:
수정하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5201,120 @@ disk_type_to_string (DB_VOLTYPE voltype)
+  int name_length_diff = 0;
+  int old_next_vol_fullname_size = 0;
+
+  assert (vhdr != NULL);
```
```suggestion
  assert (vhdr != NULL);
  assert (next_vol_fullname != NULL);
```
nit

H2SU replied:
next_vol_fullname이 NULL로 들어오는 경우가 있어 예외처리를 따로 해주고 있습니다!

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5201,120 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  int length_diff;
-  int length_to_move;
   int ret = NO_ERROR;
+  int length_to_move = 0;
```
```suggestion
```
nit

H2SU replied:
수정했습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5201,120 @@ disk_type_to_string (DB_VOLTYPE voltype)
-  if (length_diff != 0)
+  if (name_length_diff != 0)
     {
+      length_to_move = (next_vol_fullname_size + remarks_size);
```
```suggestion
      const int length_to_move = (next_vol_fullname_size + remarks_size);
```

H2SU replied:
수정했습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,18 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ * 
+ * DISK_VOLUME_HEADER = fixed fields + variable fields
+ * variable fields = volume fullname + next volume fullname + remarks
+ * volume fullname & next volume fullname max length = DB_MAX_PATH_LENGTH (include null charactor)
```
TODO 도 있으면 좋을 것 같습니다. 아래는 예시입니다. 희수님의 배경지식을 토대로 TODO를 작성해주시면 좋을 것 같네요. 문제 상황, 개선 사항, 언제 개선할지 등을 적어주시면 좋겠습니다. 물론 한글로 적으면 안됩니다. 
```suggestion
 * volume fullname & next volume fullname max length = DB_MAX_PATH_LENGTH (include null charactor)
 * TODO:
 *  DB_PAGESIZE가 16K 미만일 때는 고려되어있지 않다. 이 부분을 포함한 모든 구현부에서 고려하고 있지 않기 때문에, 추후에 DB_PAGESIZE에 따라 각 멤버의 길이를 조정하는 이슈를 진행할 때 이 계산식도 수정이 필요하다. 
```

H2SU replied:
추가하였습니다.

H2SU committed:
[CBRD-25244] Add null check and annotation, and change variable to const

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5311,16 +5364,59 @@ disk_vhdr_get_vol_remarks (const DISK_VOLUME_HEADER * vhdr)
+STATIC_INLINE int
+disk_vhdr_get_next_vol_fullname_size (const DISK_VOLUME_HEADER * vhdr)
+{
+  return (int) (strlen (disk_vhdr_get_next_vol_fullname (vhdr)) + 1);
```
vhdr 에 대하여 null 체크 하는 함수와 체크하지 않는 함수가 있는데 차이가 뭘까요?

H2SU replied:
vhdr을 인자로 받는 함수에 대해서 null체크를 전부 추가하여 수정하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5159,113 +5205,120 @@ disk_type_to_string (DB_VOLTYPE voltype)
-	{
-	  (void) strcpy (disk_vhdr_get_vol_remarks (vhdr), vol_remarks);
-	}
+      assert (remarks_size <= (int) DISK_VOLUME_HEADER_REMARKS_MAX_SIZE);
```
매크로 형변환이 필요한가요?

H2SU replied:
이 부분 빼먹고 수정을 못 한 것 같습니다. 수정 하겠습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,22 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ *
+ * TODO : The case when DB_PAGESIZE is less than 8317 bytes is not considered.
+ * This issue is not addressed in all parts of the implementation.
+ * In the future, when adjusting the lengths of each member according to DB_PAGESIZE, this calculation should also be revised.
```
제가 남긴 리뷰를 번역만 하셨네요... 명확하게 적어주셔야합니다. 어떤 멤버들이 조절되어야하는지, 어떤 경우에 문제가 생길 것인지, 등 분석한 내용이 간략하게나마 들어가야합니다. 
TODO는 개발 문맥을 잃어도, 무엇을, 언제 고쳐야하는지 쉽게 이해할 수 있어야합니다. 지금은 아무것도 알 수 없네요.
또한 comment외에 이슈로 등록해놓는게 문제상황을 추적하기 더 쉬울 것 같습니다.

H2SU replied:
TODO 코멘트 수정하였습니다. 해당 이슈는 본 이슈와 무관한 기존에도 발생할 수 있던 문제이기 때문에 다른 이슈를 생성하여 해당 이슈에서 처리 할 예정입니다!

H2SU committed:
[CBRD-25244] Add null check, change annotation, and update return type of the function.

joohok reviewed (approved):
고생하셨습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5308,19 +5367,70 @@ disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr)
+{
+  assert (vhdr != NULL);
+
+  return (size_t) (strlen (disk_vhdr_get_vol_fullname (vhdr)) + 1);
```
형변환이 필요할까요?

H2SU replied:
필요없는 것 같습니다. 제거하겠습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,24 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ * variable fields = volume fullname + next volume fullname + remarks
+ * volume fullname & next volume fullname max length = DB_MAX_PATH_LENGTH (include null charactor)
+ *
+ * TODO : The current maximum size for volume fullname is 4K.
+ * This will cause bugs if PAGESIZE is set to 4K or 8K.
+ * While the max size is 4K when PAGESIZE is 16K, it should decrease proportionally as PAGESIZE decreases.
+ * Setting the volume fullname size to 2K for an 8K PAGESIZE and 1K for a 4K PAGESIZE should work without bugs.
+ * This issue will be created and handled in a separate issue.
```
바로 이슈화 시키도록 하시죠. (별도 논의 이후)
* 관련 이슈 등록
  * databases.txt 파일의 각 라인 정보의 최대 길이가 CFG_MAX_LINE(4K)으로 인해 발생하는 이슈
  * 주석에 기술된 것처럼 4K 또는 8K 데이터 페이지 설정 시 DB_MAX_PATH_LENGTH(4K) 길이의 경로 사용하면 발생하는 이슈
* 이 PR을 develop에 반영 시 관련 이슈가 발생할 수 있습니다. 위 등록된 이슈 처리 후 같이 develop에 반영하거나 feature 브랜치에 병합하여 위 이슈랑 묶어서 develop 반영하도록 합시다. (별도 논의 이후)

H2SU replied:
넵 확인했습니다.

H2SU replied:
http://jira.cubrid.org/browse/CBRD-25444

http://jira.cubrid.org/browse/CBRD-25445

생성하였습니다.

joohok replied:
이슈번호가 잘못되었네요.

H2SU replied:
수정하였습니다.

hornetmj replied:
이슈 포맷에 맞게 작성해 주세요. 등록된 이슈 처리 시 위 주석 역시 제거되어야 합니다. 이슈에 관련 내용 역시 같이 남겨주시길 바랍니다.

hornetmj commented on src/base/error_code.h:
```diff
@@ -1739,7 +1739,9 @@
 #define ER_SP_COMPILE_ERROR                         -1360
 
-#define ER_LAST_ERROR                               -1361
+#define ER_DISK_REMARKS_LENGTH_TOO_LONG             -1361
```
```suggestion
#define ER_DISK_REMARKS_LENGTH_IS_TOO_LONG             -1361
```
ER_BO_FULL_DATABASE_NAME_IS_TOO_LONG와 일관성 있게

H2SU replied:
넵! 수정했습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,24 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ * Setting the volume fullname size to 2K for an 8K PAGESIZE and 1K for a 4K PAGESIZE should work without bugs.
+ * This issue will be created and handled in a separate issue.
+ */
+#define DISK_VOLUME_HEADER_REMARKS_MAX_SIZE (DB_PAGESIZE - (offsetof(DISK_VOLUME_HEADER, var_fields) + DB_MAX_PATH_LENGTH + DB_MAX_PATH_LENGTH))
```
수정된 코드를 보니 `offsetof(DISK_VOLUME_HEADER, var_fields)`도 자주 사용하네요. 다음과 같이 매크로화 하는게 좋겠네요.
```
#define DISK_VOLUME_HEADER_FIXED_FIELDS_SIZE (offsetof(DISK_VOLUME_HEADER, var_fields))
```

H2SU replied:
매크로화 하는게 좋아보입니다! 수정했습니다!

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +538,35 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
-  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH)
-      || (DB_PAGESIZE < (PGLENGTH) (SSIZEOF (DISK_VOLUME_HEADER) + strlen (vol_fullname) + 1)))
+  if (vol_fullname_size > DB_MAX_PATH_LENGTH ||
+      (offsetof (DISK_VOLUME_HEADER, var_fields) + vol_fullname_size) > DB_PAGESIZE)
```
NIT:

TODO 반영 시 제거되어야할 코드 입니다. 주석 남겨두시는게 좋겠습니다.

H2SU replied:
주석 추가하였습니다

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -968,10 +1006,20 @@ disk_set_link (THREAD_ENTRY * thread_p, INT16 volid, INT16 next_volid, const cha
   addr.offset = 0;
 
+  if ((next_vol_fullname_size > DB_MAX_PATH_LENGTH) ||
+      ((offsetof (DISK_VOLUME_HEADER, var_fields) + next_vol_fullname_size) > DB_PAGESIZE))
```
NIT:

TODO 반영 시 제거되어야할 코드 입니다. 주석 남겨두시는게 좋겠습니다. 추가적으로 만약 이 조건문을 남긴다면, `(offsetof (DISK_VOLUME_HEADER, var_fields) + disk_vhdr_get_vol_fullname_size(vhdr) + next_vol_fullname_size` 맞지 않을까요?

H2SU replied:
위치를 disk_get_volheader 함수 호출 이후로 변경하여 수정하겠습니다.

hornetmj replied:
답변이 이해되지 않습니다. 자세히 설명 요청드립니다.

H2SU committed:
[CBRD-25244] Add comments, mecro and remove type cast

H2SU committed:
[CBRD-25244] fix typo

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5308,19 +5371,70 @@ disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr)
+
+  const char *remarks = disk_vhdr_get_vol_remarks (vhdr);
+
+  return (size_t) ((remarks + disk_vhdr_get_vol_remarks_size (vhdr)) - ((char *) vhdr));
```
자동으로 형변환 될 것 같은데, 명시적으로 형변환해야할 이유가 있을까요?

H2SU replied:
( char* + size_t ) - char*
여서 형변환이 자동으로 size_t로 안 될 것 같아서 size_t로 형변환을 해주는 코드를 넣었습니다.

joohok replied:
자동으로 형변환 안되나요? 아니면 안될 것 같은건가요?

joohok replied:
계산 결과 음수가 될 수도 있기 때문에, 관련 처리를 하는 것으로 논의되었습니다.

H2SU replied:
음수가 되면 안 되기 때문에 assert처리하였습니다.

hornetmj commented on src/base/error_code.h:
```diff
@@ -1739,7 +1739,9 @@
 #define ER_SP_COMPILE_ERROR                         -1360
 
-#define ER_LAST_ERROR                               -1361
+#define ER_DISK_REMARKS_LENGTH_IS_TOO_LONG             -1361
```
```suggestion
#define ER_DISK_REMARKS_LENGTH_IS_TOO_LONG          -1361
```

H2SU replied:
수정하였습니다.

hornetmj commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 The length of the comment(%1$d) provided via the --comment option of createdb or addvoldb exceeds %2$d bytes.
```
첫번째 길이 정보는 언제 유용한 정보인가요? 테스트? 사용자?

H2SU replied:
사용자에게 정보를 전달할 목적으로 작성하였습니다!
지금 사용자가 작성한 comment의 길이는 몇인지를 알면 현재 comment에서 어느정도만 줄여도 가능한지 알 수 있어 도움이 될 것 같습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,26 @@ static DKNSECTS disk_Temp_max_sects = -2;
+ * 
+ * DISK_VOLUME_HEADER = fixed fields + variable fields
+ * variable fields = volume fullname + next volume fullname + remarks
+ * volume fullname & next volume fullname max length = DB_MAX_PATH_LENGTH (include null charactor)
```
```suggestion
 * volume fullname & next volume fullname max length = DB_MAX_PATH_LENGTH (include null character)
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +540,36 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
 
-  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH)
-      || (DB_PAGESIZE < (PGLENGTH) (SSIZEOF (DISK_VOLUME_HEADER) + strlen (vol_fullname) + 1)))
+  // The condition will be modified when the TODO is addressed.
```
```suggestion
  // The second condition will be modified when the TODO is addressed.
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -968,19 +1009,30 @@ disk_set_link (THREAD_ENTRY * thread_p, INT16 volid, INT16 next_volid, const cha
       return error_code;
     }
 
+  // The condition will be modified when the TODO is addressed.
```
```suggestion
  // The second condition will be modified when the TODO is addressed.
```

H2SU replied:
수정하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -516,25 +540,36 @@ disk_format (THREAD_ENTRY * thread_p, const char *dbname, VOLID volid, DBDEF_VOL
 
-  if ((strlen (vol_fullname) + 1 > DB_MAX_PATH_LENGTH)
-      || (DB_PAGESIZE < (PGLENGTH) (SSIZEOF (DISK_VOLUME_HEADER) + strlen (vol_fullname) + 1)))
+  // The condition will be modified when the TODO is addressed.
```
파일 내의 TODO는 하나 이상이 존재할 수 있습니다. 
뭐에 관련된 TODO인지, 또는 어디에 있는 TODO인지 명시해주면 좋을 것 같습니다.
아니면, 희수님이 처리할 TODO인가요?

H2SU replied:
제가 추후 처리할 TODO이긴하지만 명시해주는 것이 다른 사람이 코드를 보더라도 이해하기 쉬울 것 같아 수정하였습니다.

joohok replied:
답변이 되었기에 resolve 합니다.

H2SU committed:
[CBRD-25244] fix typo, comment and add type cast, exception

H2SU committed:
[CBRD-25244] fix typo

H2SU committed:
[CBRD-25244] fix error message

joohok commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 The length of the entered comments (%1$d) is too long. Maximum allowed is %2$d bytes.
```
어떤 comment의 길이가 긴 것인지 정보가 부족합니다. 에러메시지만으로는 에러내용을 유추하기 힘들 것 같습니다. 또한, 최대 허용 길이는 고정값이 아닌 걸로 알고 있습니다. 현 시점에서의 최대 허용길이라는 것을 명시해줬으면 좋겠습니다.

H2SU replied:
어떤 comment의 길이가 길다 라는게 무슨 뜻인지 이해하지 못 하였습니다.

joohok replied:
comment는 volume에만 존재하나요?

H2SU replied:
아하.. 이해했습니다. 문구 변경하였습니다.

joohok commented on src/storage/disk_manager.c:
```diff
@@ -968,19 +1017,30 @@ disk_set_link (THREAD_ENTRY * thread_p, INT16 volid, INT16 next_volid, const cha
 
-  /* Get the header of the previous permanent volume */
 
+
+  /* Get the header of the previous permanent volume */
```
왜 옮긴 걸까요?

H2SU replied:
수정하였습니다.

H2SU committed:
[CBRD-25244] Modification of Log Size Calculation Formula

H2SU commented:
> > > > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> > > 
> > > 
> > > 답변 부탁드립니다.
> > 
> > 
> > 예외처리는 함수 내부에서 진행하기 때문에 따로 수정은 필요 없을 것 같습니다.
> 
> 어떤 예외처리가 되어있는걸까요? 위 함수들에서 volume name 길이를 다룹니다. 이 이슈에서 volume name에 대해 logging 하는 부분들도 처리하길래 질문드렸습니다.

undo recv, redo recv 계산식 수정하였습니다.

H2SU committed:
[CBRD-25244] Add assertion

joohok commented on src/storage/disk_manager.c:
```diff
@@ -5308,19 +5379,72 @@ disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr)
+
+  const char *remarks = disk_vhdr_get_vol_remarks (vhdr);
+
+  assert (((remarks + disk_vhdr_get_vol_remarks_size (vhdr)) - ((char *) vhdr)) >= 0);
+
+  return (size_t) ((remarks + disk_vhdr_get_vol_remarks_size (vhdr)) - ((char *) vhdr));
```
```suggestion
  const int volume_header_size = remarks + disk_vhdr_get_vol_remarks_size (vhdr) - (char *) vhdr ;
  assert (volume_header_size >= 0);

  return (size_t) volume_header_size;
```

H2SU replied:
가독성 때문에 그 생각도 하긴 했었는데 상수를 추가로 선언하면 더 느려지지 않을까해서 추가하지 않았습니다.
위와 같이 작성하는 것이 더 괜찮을까요?

joohok replied:
가독성 문제도 있겠지만, 함수 호출 스택을 두번 쌓는 것보단 변수를 하나 더 선언하는게 좋을 것 같습니다.

H2SU replied:
이해됐습니다! 수정하였습니다

joohok commented:
> > > > > 아래 두 함수는 수정이 필요없을까요? RVDK_FORMAT에 관한 로그를 남길 때도 수정을 하셨길래 질문드립니다. disk_set_creation () disk_set_link ()
> > > > 
> > > > 
> > > > 답변 부탁드립니다.
> > > 
> > > 
> > > 예외처리는 함수 내부에서 진행하기 때문에 따로 수정은 필요 없을 것 같습니다.
> > 
> > 
> > 어떤 예외처리가 되어있는걸까요? 위 함수들에서 volume name 길이를 다룹니다. 이 이슈에서 volume name에 대해 logging 하는 부분들도 처리하길래 질문드렸습니다.
> 
> undo recv, redo recv 계산식 수정하였습니다.

관련하여 테스트케이스도 추가하면 좋겠습니다.

H2SU committed:
[CBRD-25244] delete blank

H2SU commented:
> 관련하여 테스트케이스도 추가하면 좋겠습니다.

log size의 차이가 생기는 부분은 gdb로 직접 size비교를 해보는 것 말곤 방법이 떠오르지 않습니다.
혹시 방법이 따로 있을까요..?

joohok commented:
> > 관련하여 테스트케이스도 추가하면 좋겠습니다.
> 
> log size의 차이가 생기는 부분은 gdb로 직접 size비교를 해보는 것 말곤 방법이 떠오르지 않습니다. 혹시 방법이 따로 있을까요..?

흠, data volume 생성 도중 서버를 죽이고 다시 살리는 케이스가 적절할 것 같습니다. 
해당 로그가 잘 남고, 로그를 이용해 복구가 가능하다는 것을 보면 좋겠네요. 
이슈 TC 리뷰하는 자리가 있을 때 다시 논의해보면 좋겠네요!

H2SU committed:
[CBRD-25244] Fix error message

H2SU committed:
[CBRD-25244] Add assertion in disk_vhdr_get_vol_header_size() function

H2SU commented:
> > > 관련하여 테스트케이스도 추가하면 좋겠습니다.
> > 
> > 
> > log size의 차이가 생기는 부분은 gdb로 직접 size비교를 해보는 것 말곤 방법이 떠오르지 않습니다. 혹시 방법이 따로 있을까요..?
> 
> 흠, data volume 생성 도중 서버를 죽이고 다시 살리는 케이스가 적절할 것 같습니다. 해당 로그가 잘 남고, 로그를 이용해 복구가 가능하다는 것을 보면 좋겠네요. 이슈 TC 리뷰하는 자리가 있을 때 다시 논의해보면 좋겠네요!

넵 알겠습니다!

joohok commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 The length of the entered volume comments (%1$d) is too long. Maximum allowed length at this page size is %2$d bytes.
```
Nit.
```suggestion
1361 The length of the comments (%1$d) for data volume is too long. Maximum allowed length at this page size is %2$d bytes.
```
at this page size가 좋을까요? 고민이네요. 
page size 라고 넣으면 page size 정보도 에러메시지에 기록되어야할텐데, 너무 많은 정보를 제공하는 것 같아서요. 
Maximum allowed size for the comment is %2$d bytes 이면 충분하겠죠? 어떻게 생각하시나요

H2SU replied:
저도 pagesize까지 전달을 해야하나 너무 많은 정보를 제공하는 것 아닌가 고민을 했었는데, 위와 같이 변경하면 괜찮을 것 같습니다.

H2SU replied:
수정하였습니다.

joohok reviewed (approved):
사소한 것만 남아서, 다시 리뷰요청 안해주셔도 됩니다! 고생하셨어요.

H2SU committed:
[CBRD-25244] Fix error message

H2SU commented:
> 사소한 것만 남아서, 다시 리뷰요청 안해주셔도 됩니다! 고생하셨어요.

엇 ... 해당 코멘트 보기 전에 re-request를 눌러버렸네용..

hornetmj commented on msg/en_US/cubrid.msg:
```diff
@@ -1447,7 +1447,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 Last Error
+1361 The length of the comment (%1$d) for data volume is too long. Maximum allowed size for the comment is %2$d bytes.
```
다른 언어셋에도 이제 추가해 주세요.

H2SU replied:
추가하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -221,6 +221,25 @@ static DKNSECTS disk_Temp_max_sects = -2;
 typedef UINT64 DISK_STAB_UNIT;
 #define DISK_STAB_UNIT_SIZE_OF sizeof (DISK_STAB_UNIT)
 
+#define DISK_VOLUME_HEADER_FIXED_FIELDS_SIZE (int) (offsetof(DISK_VOLUME_HEADER, var_fields))
```
```suggestion
#define DISK_VOLUME_HEADER_FIXED_FIELDS_SIZE ((int) offsetof(DISK_VOLUME_HEADER, var_fields))
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5308,19 +5377,73 @@ disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr)
+
+  return strlen (disk_vhdr_get_vol_remarks (vhdr)) + 1;
+}
+
+
+
+/*
```
```suggestion

/*
```

H2SU replied:
수정하였습니다.

hornetmj commented on src/storage/disk_manager.c:
```diff
@@ -5308,19 +5377,73 @@ disk_vhdr_get_next_vol_fullname (const DISK_VOLUME_HEADER * vhdr)
+  const char *remarks = disk_vhdr_get_vol_remarks (vhdr);
+  const int volume_header_size = remarks + disk_vhdr_get_vol_remarks_size (vhdr) - (char *) vhdr;
+
+  assert (volume_header_size >= 0);
```
```suggestion
  assert (volume_header_size >= 0 && volume_header_size <= DB_PAGESIZE);
```

H2SU replied:
추가하였습니다

H2SU committed:
[CBRD-25244] Add errer message for each locale

H2SU committed:
Merge branch 'feature/disk_volume_header' into feature/CBRD-25244

H2SU committed:
[CBRD-25244] Add parentheses and assert condition

H2SU committed:
Merge branch 'feature/CBRD-25244' of https://github.com/H2SU/cubrid into feature/CBRD-25244

H2SU commented:
@hornetmj 

>이슈 포맷에 맞게 작성해 주세요. 등록된 이슈 처리 시 위 주석 역시 제거되어야 합니다. 이슈에 관련 내용 역시 같이 남겨주시길 바랍니다.

이슈 포맷에 맞게 수정하였습니다. 해당 이슈 처리되는대로 주석 삭제하겠습니다.
http://jira.cubrid.org/browse/CBRD-25444
http://jira.cubrid.org/browse/CBRD-25445

joohok commented on msg/ko_KR.euckr/cubrid.msg:
diff를 통해 의도한 변경만 반영되었는지 체크하셔야합니다.

hornetmj replied:
라인 수정하시고, 저장하는 과정에서 파일 인코딩이 변경된 것 같습니다.
- 수정 전
```
$ file msg/ko_KR.euckr/cubrid.msg
msg/ko_KR.euckr/cubrid.msg: ISO-8859 text
```
- 수정 후
```
$ file ~/cubrid.msg
/home/hornetmj/cubrid.msg: UTF-8 Unicode text
```

H2SU replied:
수정하였습니다.

hornetmj commented on msg/ko_KR.utf8/cubrid.msg:
```diff
@@ -1446,7 +1446,9 @@ $ LOADDB
 1360 In line %1$d, column %2$d\nStored procedure compile error: %3$s
 
-1361 마지막 에러
+1361 데이터 볼륨의 comment 길이(%1$d)가 너무 깁니다. comment의 최대 크기는 %2$d 바이트입니다.
```
매뉴얼에서 comment는 주석으로 해석되고 있습니다.

```suggestion
1361 데이터 볼륨의 주석의 길이(%1$d)가 너무 깁니다. 주석의 최대 크기는 %2$d 바이트입니다.
```

H2SU replied:
수정하였습니다.

H2SU committed:
[CBRD-25244] Change encoding to euc_kr and modify error messages
